File: Topological Sort/LoudandRich.java

public class LoudandRich {
    public static void main(String[] args) {
        LoudandRich solution = new LoudandRich();
        int[][] richer = { { 1, 0 }, { 2, 1 }, { 3, 1 }, { 3, 7 }, { 4, 3 }, { 5, 3 }, { 6, 3 } };
        int[] quiet = { 3, 2, 5, 4, 6, 1, 7, 0 };
        int[] ans = solution.input(richer, quiet);
        System.out.println(ans);
    }

    public int[] input(int[][] richer, int[] quiet) {
        int[] quietest = new int[quiet.length];
        for (int i = 0; i < quiet.length; i++) {
            quietest[i] = searchForQuietest(richer, quiet, i, quietest);
        }
        return quietest;
    }

    public int searchForQuietest(int[][] richer, int[] quiet, int person, int[] quietest) {
        int leastQuiet = person;
        for (int i = 0; i < richer.length; i++) {
            if (richer[i][1] == person) {
                int leastQuietPerson;
                if (quietest[richer[i][0]] != 0)
                    leastQuietPerson = quietest[richer[i][0]];
                else
                    leastQuietPerson = searchForQuietest(richer, quiet, richer[i][0], quietest);
                leastQuiet = Math.min(quiet[leastQuietPerson], quiet[leastQuiet]) == quiet[leastQuiet] ? leastQuiet
                        : leastQuietPerson;
            }
        }
        return leastQuiet;
    }
}

File: Topological Sort/LoudandRich.js
var LoudandRich=function(richer,quiet){
     // The key is poorer than all the values
  const richerMap = new Array(quiet.length)
  
  for (let i = 0; i < richerMap.length; ++i) {
    richerMap[i] = []
  }
  for (const richerPair of richer) {
    richerMap[richerPair[1]].push(richerPair[0])
  }

  // lotr stands for "loudest of the richer"
  // The values are people
  const lotr = new Array(quiet.length)

  const calculateLotr = person => {
    if (lotr[person] === undefined) {
      let loudestPerson = person
      let richerLoudestPerson
      for (const richerPerson of richerMap[person]) {
        richerLoudestPerson = calculateLotr(richerPerson)
        if (quiet[richerLoudestPerson] < quiet[loudestPerson]) {
          loudestPerson = richerLoudestPerson
        }
      }
      lotr[person] = loudestPerson
    }
    return lotr[person]
  }

  for (let i = 0; i < lotr.length; ++i) {
    calculateLotr(i)
  }
  
  return lotr
}
const richer=[[1,0],[2,1],[3,1],[3,7],[4,3],[5,3],[6,3]];
const quiet=[3,2,5,4,5,6,7];
const result=LoudandRich(richer,quiet);
console.log(result);